### PR TITLE
guest can add items to cart #11

### PR DESCRIPTION
**Pull request by julyytran**
*Created join table ShopItems to link shops and their items
*Created ShopController and Shop::ItemsController
*Created shop show that is effectively an item index for all items of that shop
*Created item show that is namespaced under its shop
*Structured it in this way so that a shop has its own specific url per the spec
*All tests passing

I should have made more commits; got too into it. 

Not 100% confident on the structuring of adding the shop layer in this way, but I arrived at this way through hopping down the error trail, and it works. Thoughts? @julsfelic @Unsafepond

### CODE REVIEW

File: app/controllers/shop/items_controller.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class Shop::ItemsController < ApplicationController
+  def show
+    @item = find_item
```
Comment (julsfelic) [inline app/controllers/shop/items_controller.rb:3]: Where does the `find_item` method live? What does it do?

File: app/controllers/shops_controller.rb [hunk 2]
```diff
@@ -2,4 +2,9 @@ class ShopsController < ApplicationController
   def index
     @shops = Shop.all
   end
+
+  def show
+    @shop = Shop.find(params[:id])
```
Comment (julsfelic) [inline app/controllers/shops_controller.rb]: If you recall the one lesson we had where we wanted to find a shop by its slug? Let's say in theory we have a shop called Julian's Yummy Goods. We would end up wanting a URL like this: `/shops/julian-s-yummy-goods` There are a few steps to this, so I'll use specific code for each part. Within the `ShopsController` we want to be able to do `@shop = Shop.find_by(slug: params[:shop]`. In order to do this, we need to add a slug column to the `shops` table. Once we have that up, we can then create a callback method on the model `before_validation` (Don't remember if it is plural or not) that takes the name of the shop, uses the `pluralize` method to make a human readable URL and set that to the `slug` Below, I'll show you how the route would be setup

File: app/models/shop.rb [hunk 3]
```diff
@@ -1,2 +1,4 @@
 class Shop < ActiveRecord::Base
+  has_many :shop_items
+  has_many :items, through: :shop_items
```
Comment (julsfelic) [inline app/models/shop.rb]: In here is where we would set up the slug. Assuming we have the slug column, we could write something like this:

``` ruby
  before_validation :generate_slug

  private

  def generate_slug
    self.slug = name.parameterize
  end
```

File: app/views/shops/index.html.erb [hunk 4]
```diff
@@ -4,7 +4,7 @@
       <h1>All Shops</h1>
       <% @shops.each do |shop| %>
         <div id="shop">
-          <%= link_to shop.name, shops_path(shop) %>
+          <%= link_to shop.name, shop_path(shop) %>
```
Comment (julsfelic) [inline app/views/shops/index.html.erb]: I won't comment on each instance, but any view that is using `shop_path(shop)` would end up using `shop_path(shop: shop.slug)`

File: config/routes.rb [hunk 5]
```diff
@@ -14,7 +14,11 @@
     resources :items, only: [:new, :create, :edit, :update, :destroy]
   end
 
-  resources :shops, only: [:index]
+  resources :shops, only: [:index, :show]
+
+  namespace :shop do
```
Comment (julsfelic) [inline config/routes.rb]: So to make everything above work we need to slightly alter this route. What we could do instead is:

``` ruby
  resources :shop, path: ":shop", as: :shop do
    ...
  end
```

File: app/controllers/shop/items_controller.rb (hunk 1 above)
Comment (julyytran) [inline app/controllers/shop/items_controller.rb:3]: It lives in the application controller, as a result of some refactoring we did towards the end

def find_item
    Item.find(params[:id])
 end

File: config/routes.rb (hunk 5 above)
Comment (julyytran) [inline config/routes.rb]: Made all the slug changes, and am getting the error "No route matches {:action=>"show", :controller=>"shops", :shop=>"shop-1"} missing required keys: [:id]"

Example from class has the route
namespace :stores, path: ":store_name", as: :store do
....
end

but I also tried that and still get the same error. It is still expecting an id even though we want it to expect a slug, and where we change that expectation is in the route, yeah?

**Comment by julyytran** [comment]
Okay, finally, 
1. All tests passing!
2. Error was because it was looking for plural shops controller with item namespaced under it, but we had it as singular. Changed the route from plural to singular.
3. Changed the DB - dropped join table shop_items and put shop_id on items as primary key

Good to merge? @julsfelic @Unsafepond

**Comment by julsfelic** [comment]
Everything is looking good to me! I'll go ahead and merge it.

**Pull request merged**